Detectors/TRD/calibration/src/PadCalibCCDBBuilder.cxx

```diff
@@ -0,0 +1,551 @@
+  // mingain and maxgain (default range 0-10k)
+  if (sDetName == "")
+    sDetName = Form("hDet%i", nDet);
+  int nTrdRows = NROWC1;
+  int detInSupermodule = nDet % 30; // set the correct # of rows for stack 2
+  if (detInSupermodule >= 12 && detInSupermodule <= 17)
+    nTrdRows = NROWC0;
+  // create the 2d map to be filled
+  TH2F* hDetector = new TH2F(sDetName.Data(), sDetName.Data(), NCOLUMN, 0, NCOLUMN, nTrdRows, 0, nTrdRows);
+  // set branches of our tree
+  setTreeBranches(tree);
+  // loop over tree and fill histo
+  int nentries = tree->GetEntries();
+  for (int ientry = 0; ientry < nentries; ientry++) {
+    tree->GetEntry(ientry);
+    if ((int)mDet != nDet)
+      continue;
+    if (mChi <= 0.5 || mAmp <= 0 || mSgm <= 0 || mSgm > 1000)
+      continue; // TODO add setters to change cuts
+    if (mAdc < mingain || mAdc > maxgain)
+      continue;
+    hDetector->SetBinContent(hDetector->GetXaxis()->FindBin(mCol), hDetector->GetYaxis()->FindBin(mRow), mAdc);
+  }
+  return hDetector;
+}
+
+bool PadCalibCCDBBuilder::isHotAreaIsolated(TH2F* hDet, int column, int row, int matrixSize)
+{
+  bool isIsolated = kFALSE;
+  float averageGain = computeDetectorAverage(hDet);
+  int nSurroundingNotHot = 0;
+  int nMaxHot = TMath::Power(matrixSize + 2, 2) - TMath::Power(matrixSize, 2);
+  float averageAround = 0.;
+  int nUsedBins = 0;
+
+  int nHotOffset = 0; // offsets the nMaxHot criterion for cells at edges
+  if ((column == 1 || column == hDet->GetNbinsX()) && row > 1 && row < hDet->GetNbinsY()) {
+    nHotOffset = matrixSize + 2;
+  } else if ((row == 1 || row == hDet->GetNbinsY()) && column > 1 && column < hDet->GetNbinsX()) {
+    nHotOffset = matrixSize + 2;
+  } else if ((column == 1 || column == hDet->GetNbinsX()) && (row == 1 || row == hDet->GetNbinsY())) {
+    nHotOffset = nMaxHot / 2 + 1;
+  }
+
+  for (int i = 0; i < matrixSize + 2; i++) {
+    int icol = i - 1;
+    for (int j = 0; j < matrixSize + 2; j++) {
+      int jrow = j - 1;
+      if ((-1 < icol) && (icol < matrixSize) && (-1 < jrow) && (jrow < matrixSize))
+        // if ((abs(icol) <= (int)(matrixSize - 1) / 2) && (abs(jrow) <= (int)(matrixSize - 1) / 2))
+        continue;
+      float temp = TMath::Abs(hDet->GetBinContent(column + icol, row + jrow));
+      if (temp != 0) {
+        nUsedBins++;
+        averageAround += temp;
+      } else if (temp == 0) {
+        nSurroundingNotHot++;
+      }
+    }
+  }
+
+  if (nUsedBins > 0)
+    averageAround /= nUsedBins;
+  if (averageAround < 2 * averageGain && nSurroundingNotHot <= nHotOffset) { //<= nMaxHot){
+    isIsolated = kTRUE;
+  } else if (nSurroundingNotHot == nMaxHot) {
+    isIsolated = kTRUE;
+  }
+
+  return isIsolated;
+}
+
+int PadCalibCCDBBuilder::isolatedHotPadsContainmentSize(TH2F* hDet, int column, int row)
+{
+  int nsize = 0;
+  bool isContained = kFALSE;
+  while (!isContained) {
+    nsize++;
+    isContained = isHotAreaIsolated(hDet, column, row, nsize);
```

Comment:
Personal preference but maybe eliminate isContained? (Pretty sure the compiler drops this anyway and turns it into this)
```
do{
if (++nsize == 4)
return -1;
} while(! isHotAreaIsolated(hDet, column, row, nsize));